core-metadata/src/main/java/org/apache/kylin/measure/topn/TopNCounter.java

```diff
@@ -180,10 +200,20 @@ public void offerToHead(T item, double count) {
         }
         this.ordered = false;
 
-        this.sortAndRetain();
+        if (sortAndRetain || this.size() % this.scaleCapacity == 0) {
```

Comment:
add a condition: || this.capacity == this.scaleCapacity, it means that when kylin.engine.mr.topn-extra-sort-rate=1, it will sort.

Reply:
done